suro-kafka-producer/src/main/java/com/netflix/suro/sink/kafka/KafkaSink.java

```diff
@@ -105,34 +119,155 @@ private void setServoReporter() {
+        try {
+            producer.send(
+                new ProducerRecord(message.getRoutingKey(), partition, null, message.getMessage().getPayload()),
+                new Callback() {
+                    @Override
+                    public void onCompletion(RecordMetadata metadata, Exception e) {
+                        if (e != null) {
+                            log.error("Exception while sending", e);
+                            DynamicCounter.increment(
+                                MonitorConfig
+                                    .builder("failedRecord")
+                                    .withTag(TagKey.ROUTING_KEY, message.getRoutingKey())
+                                    .build());
+                            droppedRecords.incrementAndGet();
+                            runRecordCounterListener();
+                        } else {
+                            DynamicCounter.increment(
+                                MonitorConfig
+                                    .builder("sentRecord")
+                                    .withTag(TagKey.ROUTING_KEY, message.getRoutingKey())
+                                    .build());
+                            sentRecords.incrementAndGet();
+                            runRecordCounterListener();
+                        }
+                    }
+                });
+        } catch (Exception e) {
+            log.error("Exception before sending", e);
+            dropMessage(message);
+        }
+    }
+
+    private void dropMessage(MessageContainer message) {
+        DynamicCounter.increment(
+            MonitorConfig
+                .builder("droppedRecord")
+                .withTag(TagKey.ROUTING_KEY, message.getRoutingKey())
+                .build());
+        droppedRecords.incrementAndGet();
+        runRecordCounterListener();
     }
 
     @Override
     public void open() {
         producer = new KafkaProducer(props);
+        subscription = stream
+            .filter(new Func1<MessageContainer, Boolean>() {
+                @Override
+                public Boolean call(MessageContainer message) {
+                    if (!metadataWaitingQueuePolicy.acquire()) {
+                        dropMessage(message);
+                        return false;
+                    } else {
+                        return true;
+                    }
+                }
+            })
+            .subscribe(
+                new Action1<MessageContainer>() {
+                    @Override
+                    public void call(final MessageContainer message) {
+                        Runnable r = new Runnable() {
+
+                            @Override
+                            public void run() {
+                                try {
+                                    if (!metadataFetchedTopicSet.contains(message.getRoutingKey())) {
+                                        producer.partitionsFor(message.getRoutingKey());
```

Comment:
It won't block because it will run Runnable in the thread pool. To prevent to many Runnable backed up in the thread pool, there is counting Semaphore. If Runnables are backed up, it will drop messages.